feat(navigator): Add infrastructure for smart mission route planning (#27687)
PX4/PX4-Autopilot: docs/en/concept/mission_route_planning.md

> # Mission Route Planning Infrastructure

Reviewer: 1. Is this only for MC? I ask because a plane can't join a segment perpendicularly. 
2. What's the logic of heading to perpendicularly to the closest segment rather than joining at a waypoint (say)? Is it that you know the mission is a safe path so you always want to get to a segment and start flying, and the closest is likely going to be among the better candidates?

Author: The goal is to minimize deviations. A perpendicular projection is the closest path back to the planned route.

We can discuss how the branch-in is accepted in the PR for the smart mission join. The idea in FW mode is to join the route as quickly as possible (perpendicular) but with a nice curve by using an augmented acc rad (e.g. multiply it by 2) to reach the branch-in.

PX4 includes mission-route planning infrastructure in Navigator that provides a non-blocking full-route cache for quick access to the whole uploaded mission, and a planner that projects the vehicle and safe-point positions onto the mission path.
The goal is to minimize how far the vehicle deviates from the planned route during contingencies, for example:

- **Smart mission join**: after a GoTo or a manual reposition, project the vehicle onto every route segment and rejoin the mission at the best branch-in point.

> - **Route-following Return**: a [Return mode](../flight_modes/return.md) that follows the mission route and only branches off once it is close to the selected [safe point](../flying/plan_safety_points.md), instead of cutting straight across terrain.

Reviewer: It is usually best to start with why people might care than what is developed. I see the cache as a detail, and the route planning as why people care. In addition, only one of the things you're identifying is a contingency. Rally points are different

Mabye

```suggestion
PX4 includes mission-route planning infrastructure in Navigator for intelligently joining, following, and completing a planned route.
This infrastructure might be used for:

- **Smart mission rejoin**: Finding the best branch-in point to rejoin a mission (for example after moving to a GoTo point).
- **Route-following Return**: Choosing the best [safe point](../flying/plan_safety_points.md) exit point when using the mission route as a [Return mode](../flight_modes/return.md): branching off close to the selected point instead of cutting straight across terrain.

The infrastructure is included in the Navigator, and includes a non-blocking full-route cache for quick access to the whole uploaded mission.
It also includes a planner that projects the vehicle and safe-point positions perpendicularly onto each mission segment, then selects the best candidate using that projection (along with additional criteria such as the last segment flown).
```

Author: updated

The planner computes geometry and scoring only.
It does not publish setpoints, change the mission index, or decide when a flight mode should use the result.

> ::: info
> This page documents developer infrastructure.
> By itself it does not add any user-selectable behavior.
> :::

Reviewer: I think that this primarily documents the mission planning algorithms. I don't know from this, for example, how to access any of the behaviour, such as setting margins.

Ideally we'd link to detail on the integration points. Perhaps just say "its to come".

Perhaps:

```suggestion
::: info
This page documents the algorithms and features (such as mission cache) used for smart route planning.
At time of writing this infrastructure is not yet used.
:::
```

That last line could be replaced by something like "Information on how to use these in flight modes is covered in the respective modes." once we have implementations.

Author: updated

## Point Projection

This is the shared projection step used by both the [Vehicle Projection](#vehicle-projection) and the [Safe-Point Scoring](#safe-point-scoring) below.
The projected point is the vehicle in one case and a safe point (rally point) in the other.

The planner draws a perpendicular projection from the point onto every route segment and keeps up to three candidates.
A projection is a valid candidate only if its crosstrack distance is within a search margin of the closest candidate's crosstrack distance.
The consuming feature supplies that margin.

The example below uses a large margin.
The mission route is drawn in white and the rally-point (`R`) projections in green.
Every rally point has three projections because the margin is large.

> ![Point projection with a large search margin](../../assets/mission_route_planner/mission_route_planner_large_search_margin.png)

Reviewer: Why is one of the R drawn differently? 

It would be good to make one of these a vehicle.

The red R joins a waypoint, not a segment perpendicularly. That isn't mentioned.

Author: Added some details in a ::: details box

Author: updated the image to have a plane instead of the red "R" (it was red because selected). They probably need to be resized down once again, could you please do it @hamishwillee?

> ![Point projection with a large search margin](../../assets/mission_route_planner/mission_route_planner_large_search_margin.png)

Reviewer: Can/ should we draw the margins on all these images?

Author: I don't think there is much added value. We would need to also add all the xtrack distances as well

With a margin of zero, each rally point keeps only its single closest projection.

## Vehicle Projection

When a feature needs to rejoin the route, the first step is to project the current vehicle position onto the mission path and choose a "branch-in" point.
The vehicle may have left the route (for example with a GoTo), so the planner picks the point that best preserves mission continuity.

The branch-in point is chosen in three phases:

**Phase 1: Identifying valid candidates:**

The vehicle is projected onto the route as described in [Point Projection](#point-projection).

**Phase 2: Selecting the best branch-in point:**

The best candidate is chosen with a priority system:

- Priority 1: a candidate is selected immediately if it lies on the segment the vehicle is currently expected to be flying (the last targeted waypoint index, or the last flown `DO_JUMP` segment if the caller supplied it).
- Priority 2: if no candidate matches Priority 1, the planner scores each candidate by `A + B` and keeps the lowest:
  - A (crosstrack distance): from the vehicle to the projection on the route.

>   - B (distance to last flown segment): from the projection to whichever end of the last-flown segment is closer.

Reviewer: It might be worth adding a comment about how this works at high level. I know the examples do this, but perhaps:

> This selects the candidate that gets the vehicle back to where it was last flying if available, or the best match for a close in-sequence segment if it is not.

Author: Agreed, added

>   - B (distance to last flown segment): from the projection to whichever end of the last-flown segment is closer.

Reviewer: As below, A and B are confusing in the next section. YOu could do "the planner scores each candidate by summing the following distances, and keeps the lowest:"

Author: updated: 

- Priority 2: if no candidate matches Priority 1, the planner scores each candidate by summing the following distances, and keeps the lowest:
  - Crosstrack distance: from the vehicle to the projection on the route.
  - Distance along the route to the last-flown segment: from the projection, following the mission path to whichever end of the last-flown segment is closer.

>   - B (distance to last flown segment): from the projection to whichever end of the last-flown segment is closer.

Reviewer: - B (distance to last flown segment): from the projection to whichever end of the last-flown segment is closer.

This isn't clear - I was assuming direct flight to the end of the last segment along an arbitrary path. From the example below "plus the distance along the route back to the last-flown segment" it is actually flying the mission. 

```suggestion
  - B (distance to last flown segment): Distance from where the projection reaches the segment, following the mission path to whichever end of the last-flown segment is closer.
```

Author: updated to: 

  - A (crosstrack distance): Distance from the vehicle to the projection on the route.
  - B (distance along the route to last flown segment): Distance from the projection on the route, following the mission path to whichever end of the last-flown segment is closer.
  
  Same as suggestion but using "Distance from the projection" instead of "Distance from where the projection reaches the segment" because it's the same as point A

**Example 1:**

> The vehicle is flying an A-B-A mission, where the outbound and inbound legs are close to one another at the start.

Reviewer: I find "The vehicle is flying an A-B-A mission," to be unhelpful when ABA aren't shown on the diagram and don't necessarily mean anything. Can this be reworded or points added? In particular, we refer to A and B above, are these the same thing?

Author: removed the A-B-A, it was just meant to explain that the mission comes back to the takeoff

While on the outbound leg, the vehicles deviates east to a GoTo location (e.g. due to incoming traffic).
That GoTo location is closer to the inbound leg, but both legs are close enough that the projection step returns two valid candidates: one on the outbound leg and one on the inbound leg.
The outbound segment is the one the vehicle is currently expected to fly, so it wins immediately on Priority 1.

> Even without that rule it would still win on Priority 2, because its distance to the last-flown segment is much smaller.

Reviewer: How's that? The last flow segment (outbound) is further away from the goto point than the inbound path in the diagram

Author: Because of "B" which is the distance along the route. Updated to: 

Even without that rule it would still win on Priority 2, which measures distance _along the route_ back to the last-flown segment.
The vehicle was last flying on the outbound leg, so the outbound projection is only a short way along the route from it.
Reaching that same segment from the inbound projection would mean flying almost the entire mission back, so the inbound candidate scores far higher and loses.

For a route-following return, the vehicle follows the mission route until it reaches a "branch-off" point and then flies straight to a safe point.
The planner first finds the candidate branch-off points and then selects the safe point with the lowest total return cost.

This runs in two phases:

**Phase 1: Identifying valid candidates:**

The vehicle and all safe points are projected in a single scan of the route, using the same [Point Projection](#point-projection) described above.

**Phase 2: Selecting the best projection point:**

Each candidate branch-off is scored by total path cost, and the lowest wins. The cost is built from:

- Along-route distance: along the route geometry from the vehicle projection to the safe-point projection (branch-off point), using straight lines between waypoints.
- Branch-off leg: the straight-line distance from the safe-point projection to the safe point (the off-route leg flown after leaving the route).
- U-turn penalty: for fixed-wing and VTOL-in-FW, an extra distance penalty is added when the path would require an immediate U-turn, so forward-flowing paths are preferred. The caller supplies the penalty value (set it to 0 to disable).

> Safe points are filtered while the planner builds its batch:

Reviewer: What does "Builds its batch" mean?

Author: clarified: 

As the planner gathers the safe points to consider, it filters out the ones it cannot use:

- U-turn penalty: for fixed-wing and VTOL-in-FW, an extra distance penalty is added when the path would require an immediate U-turn, so forward-flowing paths are preferred. The U-turn is detected by comparing the vehicle's current velocity with the direction it would have to fly along the route toward the goal: if the two are more than 90° apart, the vehicle would have to turn around, and the penalty is applied. The caller supplies the penalty value (set it to 0 to disable). Multirotors and hovering VTOL are exempt, because they can turn on the spot.

As the planner gathers the safe points to consider, it filters out the ones it cannot use:

- Safe points are read through the provider. `MissionRouteCache` serves only a fully loaded RAM generation, so planning never reads Dataman directly.
- Invalid coordinates, unsupported frames, or filtered safe points are skipped.
- Every remaining safe point gets up to three projections.

Safe-point branch-off candidates on loop segments are more restricted:

- If the vehicle is not itself flying a loop, safe-point projections on loop segments are skipped.

>   This prevents a return from entering a mission loop just to reach a branch-off point.

Author: Add a comment explaining that this allows to simplify the overall behavior: no need to handle how to get onto the loop jump segment
